File: lib/Aoc2018/Day01.ex
defmodule Aoc2018.Day01 do
  def part_one(input) do
    input
    |> parse
    |> Enum.sum()
  end

  def part_two(input) do
    input
    |> parse
    |> process(MapSet.new([0]), [], 0)
  end

  defp parse(input) do
    input
    |> String.split("\n")
    |> Enum.map(fn x -> String.to_integer(x) end)
  end

  def process(input, frequencies, [], freq), do: process(input, frequencies, input, freq)

  def process(input, frequencies, [inst | tail], freq) do
    freq = freq + inst

    cond do
      MapSet.member?(frequencies, freq) -> freq
      true -> process(input, MapSet.put(frequencies, freq), tail, freq)
    end
  end
end

File: 2018/ruby/lib/aoc2018/day01.rb
require 'set'

module Aoc2018
  class Day01
    def self.part_one(input)
      input.split("\n").map(&:to_i).sum
    end

    def self.part_two(input)
      freqs = Set[]
      current = 0
      i = 0
      input = input.split("\n").map(&:to_i)
      while freqs.add?(current)
        current += input[i % input.length]
        i += 1
      end
      current
    end
  end
end
